File: Leetcode_with_py/200numIslands.py
class Solution(object):
    directions = [(-1, 0), (0, 1), (1, 0), (0, -1)]

    def dfs(self, grid, visited, m, n, x, y):
        visited[x][y] = True
        for direction in self.directions:
            newX = x + direction[0]
            newY = y + direction[1]
            if 0 <= newX < m and 0 <= newY < n and not visited[newX][newY] and \
                    grid[newX][newY] == '1':
                self.dfs(grid, visited, m, n, newX, newY)

    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        m = len(grid)
        if m == 0:
            return 0
        n = len(grid[0])
        visited = [[False for _ in range(n)] for _ in range(m)]
        res = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == '1' and not visited[i][j]:
                    res += 1
                    self.dfs(grid, visited, m, n, i, j)
        return res

File: Leetcode_with_Cpp/200numIslands.cpp
class Solution {
private:
    int d[4][2] = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};
    int m, n;
    vector<vector<bool>> visited;
    bool isArea(int x, int y) {
        return x >= 0 && x < m && y >=0 && y < n;
    }
    // 从 grid[x][y] 位置开始，进行 floodfill
    void dfs(vector<vector<char>>& grid, int x, int y) {
        visited[x][y] = true;
        for (int i = 0; i < 4; i++) {
            int newX = x + d[i][0];
            int newY = y + d[i][1];
            if (isArea(newX, newY) && !visited[newX][newY] && grid[newX][newY] == '1')
                dfs(grid, newX, newY);
        }
        return;
    }
public:
    int numIslands(vector<vector<char>>& grid) {
        m = grid.size();
        if (m == 0)
            return 0;
        n = grid[0].size();
        visited = vector<vector<bool>>(m, vector<bool>(n, false));
        int res = 0;
        for (int i = 0; i < m; i++)
            for (int j = 0; j < n; j++)
                if (grid[i][j] == '1' && !visited[i][j]) {
                    res++;
                    dfs(grid, i, j);
                }
        return res;
    }
};